app/uk/gov/hmrc/gform/sharedmodel/formtemplate/FormatExpr.scala

```diff
@@ -38,25 +40,77 @@ object DateConstraintType {
   implicit val format: OFormat[DateConstraintType] = derived.oformat[DateConstraintType]
 }
 
-final case class DateConstraint(beforeOrAfter: BeforeOrAfter, dateFormat: DateConstraintInfo, offset: OffsetDate)
+final case class DateConstraint(
+  beforeAfterPrecisely: BeforeAfterPrecisely,
+  dateFormat: DateConstraintInfo,
+  offset: OffsetDate)
 
 object DateConstraint {
   implicit val format: OFormat[DateConstraint] = derived.oformat[DateConstraint]
 }
 
-sealed trait BeforeOrAfter
-case object After extends BeforeOrAfter
-case object Before extends BeforeOrAfter
+sealed trait BeforeAfterPrecisely
+case object After extends BeforeAfterPrecisely
+case object Before extends BeforeAfterPrecisely
+case object Precisely extends BeforeAfterPrecisely
+
+object BeforeAfterPrecisely {
+  implicit val format: OFormat[BeforeAfterPrecisely] = derived.oformat[BeforeAfterPrecisely]
+}
+
+sealed trait DateParameters
+
+sealed trait Year extends DateParameters
+case object Next extends Year
+case object Previous extends Year
+case object AnyYear extends Year
+case class ExactYear(year: Int) extends Year
+
+object Year {
+  implicit val format: OFormat[Year] = derived.oformat[Year]
+}
+
+sealed trait Month extends DateParameters
+case object AnyMonth extends Month
+case class ExactMonth(month: Int) extends Month
 
-object BeforeOrAfter {
-  implicit val format: OFormat[BeforeOrAfter] = derived.oformat[BeforeOrAfter]
+object Month {
+  implicit val format: OFormat[Month] = derived.oformat[Month]
+}
+
+sealed trait Day extends DateParameters
+case object AnyDay extends Day
+case class ExactDay(day: Int) extends Day
+case object FirstDay extends Day
+case object LastDay extends Day
+
+object Day {
+  implicit val format: OFormat[Day] = derived.oformat[Day]
 }
 
 sealed trait DateConstraintInfo
 case object Today extends DateConstraintInfo
-case class ConcreteDate(year: Int, month: Int, day: Int) extends DateConstraintInfo
-case class NextDate(month: Int, day: Int) extends DateConstraintInfo
-case class PreviousDate(month: Int, day: Int) extends DateConstraintInfo
+
+case class ConcreteDate(year: Year, month: Month, day: Day) extends DateConstraintInfo {
+
+  val isExact: Boolean = (year.isInstanceOf[ExactYear] || year == Next || year == Previous) && month
```

Comment:
Can we pattern match, or have isExact functions on Year, Month & Day, rather than using isInstanceOf?